synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -298,10 +302,19 @@ async def action_for_event_by_user(
             if uid in ignorers:
                 continue
 
-            display_name = None
-            profile_info = room_members.get(uid)
-            if profile_info:
-                display_name = profile_info.display_name
+            # This is a check for the case where user joins a room without being
+            # allowed to see history, and then the server receives a delayed
+            # event from before the user joined, which they should not be pushed
+            # for
+            visible = await filter_events_for_client_with_state(
+                self.store, uid, event, context
+            )
+            if not visible:
+                continue
+
+            localpart = get_localpart_from_id(uid)
+            profile_info = await self.store.get_profileinfo(localpart)
```

Comment:
This gets the display name from the profile table (i.e. the users global profile), rather than the per room profile. It's also not cached.

I don't think we have an existing DB function for getting per room profiles, but I think we can join `current_state_events` and `room_memberships` to get the display name out.

We should also make this a batched look up before the loop.

Reply:
Any chance I could use the public version of this function: https://github.com/matrix-org/synapse/blob/13e359aec8ae8be8dc56a036ae6d9f2bc1d07385/synapse/storage/databases/main/roommember.py#L233
which seems to pull per-room profiles?

Reply:
So yeah, that returns the right information! The only problem with that is that it's pulling out profiles for *all* users, and ideally we're trying to reduce the amount of data we're pulling out into memory.

What we probably want is that same query, but that accepts a list of user IDs to query, using `make_in_list_sql_clause`:

https://github.com/matrix-org/synapse/blob/13e359aec8ae8be8dc56a036ae6d9f2bc1d07385/synapse/storage/databases/main/roommember.py#L426-L428